File: ruby/easy/construct_string_from_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} t
# @return {String}
def tree2str(t)
    return "" unless t
    str = t.val.to_s
    left = tree2str(t.left)
    right = tree2str(t.right)
    if(!left.empty? && !right.empty?)
        str += "(#{left})(#{right})"
    elsif(!left.empty? && right.empty?)
        str += "(#{left})"
    elsif(left.empty? && !right.empty?)
        str += "()(#{right})"
    end
    return str
end

File: javascript/easy/constructStringFromBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} t
 * @return {string}
 */
var tree2str = function(t) {
    if(t === null) return "";
    if(t.left === null && t.right === null) {
        return t.val.toString();
    }
    let str = t.val + "(" +tree2str(t.left) + ")";
    let right = tree2str(t.right);
    if(right !== "") {
        str += "(" + right + ")";
    }
    return str;
};
